Add to ros2 documentation one or multiple tutorials to create a first simulation with Webots

BenjaminHug8 opened the pull request:
**Context**
This PR aims to create a tutorial on the [official ros2 documentation page](https://docs.ros.org/en/rolling/Tutorials.html) most probably in the **Advanced** tab.

There is already our tutorials with `Turtlebot 3` in the **Demos** section, but this is not as visible as a tutorial.

This should give more information to a new user of Webots how to use ROS2 with the simulator.

**Tips**
To test the tutorial, follow the instructions at the top of this [page](https://docs.ros.org/en/rolling/Contributing/Contributing-To-ROS-2-Documentation.html).

Once it is ready to be published in a PR on the official repository:

- [x] Remove extra images not used.
- [x] Remove `Tutorial-Create-Webots-Robot.md` as it is here only to be reviewed.

BenjaminHug8 commented:
I am still correcting some misspelling but I would be glad if you can take a first look @omichel .

omichel commented:
I would rename the title to something more in line with other tutorials, like "Running a Robot Simulation (Webots)".

omichel commented on source/Tutorials/Simulators/Creating-Your-First-ROS2-Webots-Simulation.rst:
```diff
@@ -0,0 +1,451 @@
+Background
+----------
+
+Webots is an open source and multi-platform robot simulator.
+It provides a complete development environment to model, program and simulate robots.
```
Let's put it in context of the previous simulator used, e.g., turtlesim:
```suggestion
Webots is a robot simulator that can used with ROS 2.
Unlike turtlesim, it provides physics-based models for robots, sensors, actuators and objects.
This provides realistic simulation results that should match the behaviors observed in the real world.
Hence, what you get in simulation is very close to what you will observe when transferring your ROS 2 controllers to a real robot.
```

omichel commented on source/Tutorials/Simulators/Creating-Your-First-ROS2-Webots-Simulation.rst:
```diff
@@ -0,0 +1,451 @@
+Webots is an open source and multi-platform robot simulator.
+It provides a complete development environment to model, program and simulate robots.
+
+Webots_ros2 is a package offering an interface between ROS2 and Webots, using the sub-package named webots_ros2_driver that you will use in this tutorial.
```
```suggestion
webots_ros2 is a package offering an interface between ROS2 and Webots, using the sub-package named webots_ros2_driver that you will use in this tutorial.
```

omichel commented on source/Tutorials/Simulators/Creating-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,332 @@
+Creating a Robot Simulation (Webots)
+====================================
+
+**Goal:** Install and discover the Webots simulator with the webots_ros2 package.
```
```suggestion
**Goal:** Setup a robot simulation and control it from ROS 2.
```

omichel commented on source/Tutorials/Simulators/Creating-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,332 @@
+.. _Simulators:
+
+Creating a Robot Simulation (Webots)
```
I would change "Creating" to "Setting-up" or "Running" because we not creating the simulation world file in there.

omichel commented on source/Tutorials/Simulators/Creating-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,332 @@
+Hence, what you get in simulation is very close to what you will observe when transferring your ROS 2 controllers to a real robot.
+
+webots_ros2 is a package offering an interface between ROS2 and Webots with the help of the sub-package webots_ros2_driver that you will use in this tutorial.
+There is also some other sub-packages containing examples with different robots such as the TurtleBot 3 you may find on the :doc:`../../Tutorials` page under the ``Demos`` section.
```
```suggestion
There are also some other sub-packages containing demos with different robots such as the TurtleBot 3.
Such demos are detailed in the :doc:`../../Tutorials` page under the ``Demos`` section.
```

omichel commented on source/Tutorials/Simulators/Creating-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,332 @@
+webots_ros2 is a package offering an interface between ROS2 and Webots with the help of the sub-package webots_ros2_driver that you will use in this tutorial.
+There is also some other sub-packages containing examples with different robots such as the TurtleBot 3 you may find on the :doc:`../../Tutorials` page under the ``Demos`` section.
+
+This tutorial will introduce you how to create your first simulation by modeling your robot and implementing a Python plugin derived from webots_ros2_driver.
```
We not any more modeling the robot here.

omichel commented on source/Tutorials/Simulators/Creating-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,332 @@
+Precedent tutorials
+^^^^^^^^^^^^^^^^^^^
+
+This tutorial assume you followed the tutorials :doc:`../Creating-Your-First-ROS2-Package` and :doc:`../Launch-Files/Using-ROS2-Launch-For-Large-Projects`.
```
```suggestion
This tutorial assumes you followed the tutorials :doc:`../Creating-Your-First-ROS2-Package` and :doc:`../Launch-Files/Using-ROS2-Launch-For-Large-Projects`.
```

omichel commented on source/Tutorials/Simulators/Creating-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,332 @@
+Prerequisites
+-------------
+
+Precedent tutorials
```
```suggestion
Previous tutorials
```

omichel commented on source/Tutorials/Simulators/Creating-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,332 @@
+Summary
+-------
+
+In this tutorial, you created a simulation with Webots, modeled your own robot and implemented a Python plugin to use its motors and sensors.
```
We didn't modeled anything.

omichel commented on source/Tutorials/Simulators/Creating-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,332 @@
+              ├── setup.cfg
+              └── setup.py
+
+2 Create the simulation world in Webots
```
We don't really create the simulation here.
```suggestion
2 Setup the simulation world in Webots
```

lukicdarkoo commented on source/Tutorials/Simulators/Code/my_robot_driver.py:
```diff
@@ -0,0 +1,53 @@
+import math
```
This is very nice, but we can expose the same interface through the URDF configuration file with `webots_ros2_control`. Why did you choose to go with this approach?

BenjaminHug8 replied:
The idea was to not use `webots_ros2_control` as it depends on `ros2_control` and we wanted to keep it with the minimum amount of packages (only 4 packages here).

omichel replied:
Yes, that could be the next tutorial indeed...

BenjaminHug8 replied:
It does not cost much to write a note, I ll do it then!

lukicdarkoo reviewed (changes requested):
It looks very good and clean. However, I would rather simplify the tutorial and allow a user to control the robot (with `cmd_vel`) as soon as possible with as least code as possible. Also, instead of the Webots robot, I would rather go with a TurtleBot to depict Webots' visual fidelity. What do you think?

BenjaminHug8 commented:
> It looks very good and clean. However, I would rather simplify the tutorial and allow a user to control the robot (with `cmd_vel`) as soon as possible with as least code as possible. Also, instead of the Webots robot, I would rather go with a TurtleBot to depict Webots' visual fidelity. What do you think?

The idea was to have a tutorial where the user starts from nothing and then create his own robot. With Olivier we thought that keeping the part where the user build his robot would not be accepted on the official ROS2 documentation, thus I put it on the Gihub of `webots_ros2`.
On one hand I think that using Turtlebot will be good looking and somewhat will link well with the 2 demos we have, on the other hand I also think users will benefit of a tutorial on how building their robot.

omichel commented:
Yes, one of the advantages of using a super simple robot in a world file is that it is simpler for the beginners: a single world file, no concept of proto, the devices of the robot are minimalist and can be easily found in the world file, which is super small comparing to a Turtlebot proto.

BenjaminHug8 commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,333 @@
+
+This tutorial assumes you followed the tutorials :doc:`../Creating-Your-First-ROS2-Package` and :doc:`../Launch-Files/Using-ROS2-Launch-For-Large-Projects`.
+
+Install Webots
+^^^^^^^^^^^^^^
+
+Go on the `Webots Github release page <https://github.com/cyberbotics/webots/releases/latest>`_ to download the Debian package and then run the following command, ``VERSION`` being the version of Webots you downloaded:
+
+.. code-block:: bash
+
+        sudo apt install ./webots_VERSION_amd64.deb
```
That is true. But in the case the user will build his robot or just want to open the world given to see what it looks like I think it will be great he can install it (before launching any `webots_ros2` related stuff)?

But a comment to indicate it can be automatically downloaded is good tho.

omichel replied:
I tend to agree with @lukicdarkoo: if not needed, this will simplify the instructions and we don't expect the user will do anything else than what (s)he is instructed during the tutorial, so I believe installing Webots explicitly is superfluous here.

BenjaminHug8 replied:
Ok I will change that.

BenjaminHug8 commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,333 @@
+
+Simulations in Webots rely on world files containing objects called ``nodes`` (different from the standard ROS node).
+You will need a world file with a robot to launch your simulation.
+Please go on this `page <https://github.com/cyberbotics/webots_ros2/wiki/Tutorial-Create-Webots-Robot>`_ to learn how to create your own robot or download a premade one.
```
It is a good point to have something quickly runable for the user to test. Is the fact to go on our Github to download the file of the world the same that recommend them to import a premade robot ?

omichel replied:
Instead of inviting the users to another tutorial here, we should point directly to the zip file to download, e.g., https://github.com/cyberbotics/webots_ros2/wiki/assets/tutorialCreateWebotsRobot/my_world.zip and refer to the tutorial afterwards as an option in case the user want to understand the details of this file.

BenjaminHug8 commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,333 @@
+
+Return to the file ``my_package/my_package/my_robot_driver.py`` and replace the code inside with:
+
+.. literalinclude:: Code/my_robot_driver_sensor.py
```
You mean keep the first `my_robot_driver.py` that is listening `/cmd_vel` and just create a node that listen to the sensors and publish commands on `/cmd_vel`?

Don't you think that it would add a little bit more complexity as there is more files and more nodes to launch? (Thus an extra node is more clean if we want to keep the behavior of the first `my_robot_driver.py` , I admit).

lukicdarkoo replied:
> You mean keep the first my_robot_driver.py that is listening /cmd_vel and just create a node that listen to the sensors and publish commands on /cmd_vel?

Exactly.

> Don't you think that it would add a little bit more complexity as there is more files and more nodes to launch? (Thus an extra node is more clean if we want to keep the behavior of the first my_robot_driver.py , I admit).

There is the same number of files and you have only one extra node to launch. Although it is slightly simpler to use Webots ROS 2 plugins it is generally a bad practice. The obstacle avoidance controller should work with a real robot as well.

BenjaminHug8 replied:
I will change that.

BenjaminHug8 commented:
I handled the comments of @lukicdarkoo in the last commit.

I will wait the comments of Joachim.

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Your robot will use a standard ROS node in order to detect the wall and send commands to avoid it.
+Add a file ``obstacle_avoider.py`` to ``my_package/my_package/`` and fill it with:
```
consistency (in .. create a file... )

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+.. literalinclude:: Code/obstacle_avoider.py
+    :language: python
+
+This node will simply create a publisher for the command and subscribe to the sensors topics here:
```
```suggestion
This node will create a publisher for the command and subscribe to the sensors topics here:
```

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+
+And finally when a measure is recieved from the right sensor a command will be sent to the topic ``/cmd_vel`` in case any of the two sensors has detect an obstacle.
+If the right sensor detects an obstacle the command will make the robot turn in place clockwise.
+Otherwise if only the left sensor detects an obstacle the command will make the robot turn right.
```
```suggestion
Otherwise, if only the left sensor detects an obstacle the command will make the robot turn anticlockwise.
```

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+    :language: python
+    :lines: 19-20
+
+And finally when a measure is recieved from the right sensor a command will be sent to the topic ``/cmd_vel`` in case any of the two sensors has detect an obstacle.
```
```suggestion
Finally, a command will be sent to the topic ``/cmd_vel`` when a measure is received from a sensor, i.e. when one of the two sensors has detected an obstacle.
```

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+12 Updating setup.py and robot_launch.py
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You will also have to slightly modify these two other files in order to launch your new node.
```
```suggestion
You have to modify these two other files to launch your new node.
```

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+    :language: python
+    :lines: 24-27
+
+And finally in your file ``robot_launch.py`` replace ``def generate_launch_description():`` by the code bellow:
```
consistency

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+    :language: python
+    :lines: 10-42
+
+This will create an ``obstacle_avoider`` node and it will be prepared to be launched in ``LaunchDescription``.
```
```suggestion
This will create an ``obstacle_avoider`` node. It will be prepared to be launched in ``LaunchDescription``.
```
Sentence a bit long, not clear but don't know how to do better.

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You will also have to slightly modify these two other files in order to launch your new node.
+Go to your file ``setup.py`` and add ``'obstacle_avoider = my_package.obstacle_avoider:main'`` to ``'console_scripts'``:
```
consistency

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+
+.. image:: Image/Step_25.png
+
+9 Going further
```
The `going further` part actually includes all the following

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Repeat the same commands as in tasks ``8`` to test your code.
+From a terminal in your ROS2 working directory run:
```
```suggestion
From a terminal in your ROS2 workspace run:
```

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+
+            ros2 topic pub /cmd_vel geometry_msgs/Twist  '{linear:  {x: 0.1, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: 0.0}}'
+
+First your robot will go forward and before hitting the wall it will turn clockwise.
```
```suggestion
Your robot will go forward and before hitting the wall it will turn clockwise.
```

joachimhgg reviewed (commented):
Feel free to take in account or not the suggestions, some of them are obvious or subjective. Usually for the ROS tutorials, does it has to be code and then explication or code following explication? In any case I think it would be better to keep the same order.
I found it makes the tutorial heavier the fact to duplicate some part of the code to explain them.
Otherwise, the tutorial is really clear and easy to follow, I did not had any issue to do or understand it..

BenjaminHug8 commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,366 @@
+
+This primarly modificate the code in two places.
+
+.. literalinclude:: Code/setup.py
```
I agree for the `Here`.

Otherwise concerning code and explications it seems to not be hard constraint. [Here](https://docs.ros.org/en/rolling/Tutorials/Writing-A-Simple-Py-Publisher-And-Subscriber.html#examine-the-code) is an example.

But generally the explanations come before the code so I will adapt this.

BenjaminHug8 commented:
> I found it makes the tutorial heavier the fact to duplicate some part of the code to explain them.

If you check the other ROS tutorials, you will see that usually when there is a file they will first show it entirely and then repeat the code block by block to explain it.
But it is true that if I can remove some repetition it will be lighter to read.

BenjaminHug8 commented:
I integrated the review of @joachimhgg in the 2 last commits.

Let me know if you have any comments or suggestions about this tutorial (and the one on [webots_ros2](https://github.com/cyberbotics/webots_ros2/wiki/Tutorial-Create-Webots-Robot#build-the-robot).

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,352 @@
+Simulations in Webots rely on world files containing objects called ``nodes`` (different from the standard ROS node).
+You will need a world file with a robot to launch your simulation.
+Please `download this world file <https://github.com/cyberbotics/webots_ros2/wiki/assets/tutorialCreateWebotsRobot/my_world.zip>`_ and extract it in the folder ``my_package/worlds/``.
+A custom robot is already modeled in this world ``my_world.wbt``.
```
```suggestion
This is actually a fairly simple text file you can visualize in a text editor.
A simple robot is already included in this ``my_world.wbt`` world file.
```

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,352 @@
+3 Prepare the package.xml file
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the ``<package format="3">`` balise add the following:
```
```suggestion
Add the following inside the ``<package format="3">`` tag:
```

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,352 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You will start by modify these two files in order to enable the sensors.
+Go to your file ``package.xml`` and in the ``<package format="3">`` balise add the following:
```
```suggestion
Go to your file ``package.xml`` and in the ``<package format="3">`` tag add the following:
```

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,352 @@
+
+The ROS2 interface will use the standard parameters in the ``<ros>`` tags to enable the **DistanceSensor** nodes and name their topics.
+
+10 Creating a ROS node to avoid obstacle
```
```suggestion
10 Creating a ROS node to avoid obstacles
```

BenjaminHug8 commented:
I added the file `Tutorial-Create-Webots-Robot.md`. This file is the tutorial on the `webots_ros2` on how to construct a robot in webots.

In order to make reviews on it I put a copy in this PR.

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+
+![cylinder_DEF](./assets/tutorialCreateWebotsRobot/cylinder_DEF.png)
+
+Add to your robot a `Physics` node to the corresponding field in order to enable the collision with other objects.
```
```suggestion
Add to your robot a `Physics` node to the field `physics` in order to enable the collision with other objects.
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+
+![hinge_joint](./assets/tutorialCreateWebotsRobot/hinge_joint.png)
+
+Then select `jointParameters` and add a `HingeJointParameters` node.
```
```suggestion
Then, select `jointParameters` and add a `HingeJointParameters` node.
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+![hinge_joint](./assets/tutorialCreateWebotsRobot/hinge_joint.png)
+
+Then select `jointParameters` and add a `HingeJointParameters` node.
+Select `device`, add a `RotationalMotor` node and then configure the `HingeJointParameters` and `RotationalMotor` nodes like below.
```
```suggestion
Select `device`, add a `RotationalMotor` node and configure the `HingeJointParameters` and `RotationalMotor` nodes like below.
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+
+![left_motor](./assets/tutorialCreateWebotsRobot/left_motor.png)
+
+`axis` will indicate around which axis the wheel should rotate and `anchor` indicates an offset.
```
```suggestion
`axis` will indicate around which axis the wheel should rotate and `anchor` indicates an offset.
```
which offset?

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+![left_motor](./assets/tutorialCreateWebotsRobot/left_motor.png)
+
+`axis` will indicate around which axis the wheel should rotate and `anchor` indicates an offset.
+Next select `endPoint`, add a `Solid` node and set it up like here in a similar manner you did for the body of the robot:
```
```suggestion
Select `endPoint`, add a `Solid` node and set it up like here in a similar manner you did for the body of the robot:
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+
+![left_wheel_solid](./assets/tutorialCreateWebotsRobot/left_wheel_solid.png)
+
+Then change the name to `left wheel` and add to the `boundingObject` a `WHEEL` node (you will find it again in the `USE` field).
```
```suggestion
Change the name to `left wheel` and add to the `boundingObject` a `WHEEL` node (you will find it again in the `USE` field).
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+![left_wheel_solid](./assets/tutorialCreateWebotsRobot/left_wheel_solid.png)
+
+Then change the name to `left wheel` and add to the `boundingObject` a `WHEEL` node (you will find it again in the `USE` field).
+Add to your robot a `Physics` node to its corresponding field once again in order to enable the collisions.
```
```suggestion
Add to your `endPoint Solid` a `Physics` node to its field `physics` once again to enable the collisions.
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+
+### Right wheel of the robot
+
+Creating the second wheel is bit easier as you can use the `DEF-USE` mechanism of Webots.
```
```suggestion
Creating the second wheel is easier as you can use the `DEF-USE` mechanism of Webots.
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+### Right wheel of the robot
+
+Creating the second wheel is bit easier as you can use the `DEF-USE` mechanism of Webots.
+Select `Robot` > `children` > `HingeJoint` and then add a `HingeJoint` node.
```
```suggestion
Select `Robot` > `children` > `HingeJoint` and add a `HingeJoint` node.
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+
+Let's add some sensors in order to make the robot able to detect front obstacles.
+Start by adding a `DistanceSensor` node in the `children` field of your robot.
+Then configure the sensor as described in the following image in order to correctly place and model the sensor:
```
```suggestion
Then, configure the sensor as described in the following image in order to correctly place and model the sensor:
```

joachimhgg commented on source/Tutorials/Simulators/Tutorial-Create-Webots-Robot.md:
```diff
@@ -0,0 +1,165 @@
+
+![Sensor_1_2](./assets/tutorialCreateWebotsRobot/Sensor_1_2.png)
+
+Creating the second sensor is straightforward and you can follow this structure:
```
```suggestion
Creating the second sensor is straightforward and you can copy/paste the `ds0 DistanceSensor `node and follow this structure:
```

joachimhgg reviewed (commented):
About the tutorial to create a robot on `webots_ros2`:

It is an easy tutorial but redundant and long to do since you have to do all your robot by hand. Nevertheless, I think it is usefull when you 're starting with Webots and it is well explained. I compared the file that I have done thanks to the tutorial with the given world and it is not the same. I sent it to you. 

Maybe we can explain more how to run rosrun at the end? don't know if it is the purpose of the tutorial

BenjaminHug8 commented:
> Maybe we can explain more how to run rosrun at the end? don't know if it is the purpose of the tutorial

For me we can expect users to know some basis of ROS for this tutorial.

omichel commented:
> For me we can expect users to know some basis of ROS for this tutorial.

Yes, if some knowledge from other basic ROS tutorials is needed, we should mention them in the prerequisites section.

BenjaminHug8 commented:
> I compared the file that I have done thanks to the tutorial with the given world and it is not the same.

One mistake was due to a wrong image in the tutorial, otherwise it is quite correct. I think it will be normal for users to make mistakes and to compare the given world with their.

BenjaminHug8 commented:
@omichel may I ask you to test the 2 tutorials?

In case you want to make a review on the tutorial `Create Webots Robot` which is on [webots_ros2](https://github.com/cyberbotics/webots_ros2/wiki/Tutorial-Create-Webots-Robot), there is a copy of the .md file in this PR.

lukicdarkoo commented on source/Tutorials/Simulators/Code/obstacle_avoider.py:
```diff
@@ -0,0 +1,52 @@
+
+        self.__publisher = self.create_publisher(Twist, 'cmd_vel', 1)
+
+        self.create_subscription(Range, 'l_sensor', self.__l_sensor_callback, 1)
+        self.create_subscription(Range, 'r_sensor', self.__r_sensor_callback, 1)
```
```suggestion
        self.create_subscription(Range, 'left_sensor', self.__left_sensor_callback, 1)
        self.create_subscription(Range, 'right_sensor', self.__right_sensor_callback, 1)
```

lukicdarkoo commented on source/Tutorials/Simulators/Code/obstacle_avoider.py:
```diff
@@ -0,0 +1,52 @@
+        self.create_subscription(Range, 'l_sensor', self.__l_sensor_callback, 1)
+        self.create_subscription(Range, 'r_sensor', self.__r_sensor_callback, 1)
+
+    def __l_sensor_callback(self, msg):
+        self.__l_sensor_value = msg.range
+
+    def __r_sensor_callback(self, msg):
+        self.__r_sensor_value = msg.range
```
```suggestion
    def __left_sensor_callback(self, message):
        self.__left_sensor_value = message.range

    def __right_sensor_callback(self, message):
        self.__right_sensor_value = message.range
```

omichel commented:
I believe we need to handle the Linux / Windows / macOS cases for command lines instructions like installing the webots_ros2 package, creating directories, etc., like it is done in other tutorials, for example https://docs.ros.org/en/foxy/Tutorials/Topics/Topic-Statistics-Tutorial.html

lukicdarkoo commented:
> I believe we need to handle the Linux / Windows / macOS cases for command lines instructions like installing the webots_ros2 package, creating directories, etc., like it is done in other tutorials, for example https://docs.ros.org/en/galactic/Tutorials/FastDDS-Configuration/FastDDS-Configuration.html and https://docs.ros.org/en/foxy/Tutorials/Topics/Topic-Statistics-Tutorial.html

For Windows, we need to handle this first:
https://github.com/cyberbotics/webots_ros2/issues/303#issuecomment-976574206

Regarding macOS, I don't think it is relevant anymore as the support has been downgraded to Tier 3.

omichel commented:
I made many small modifications, fixed English, fixed details, added references, provided better explanations, etc.
The first part (until the robot moves forwards) is now pretty satisfactory to me.
However, the second parts involving the sensors seems weird: (1) it doesn't work and (2) it seems we are sending contradictory twist messages at the same time (which could explain why it doesn't work). I believe we should provide a cleaner example.

BenjaminHug8 commented:
> However, the second parts involving the sensors seems weird

I will check if I can simplify some parts.

> (1) it doesn't work

A change in the name of the variables broke a part, it works on my side now.

> (2) it seems we are sending contradictory twist messages at the same time

This part is also solved.
I pushed a commit for 1) and 2).

omichel commented:
OK, it looks good to me now.
I changed it quite a bit since I started the review.
So I believe it would be nice if @lukicdarkoo and @joachimhgg could go through it carefully once again to be sure it looks all good also for you.
The goal is really to show the most simple system exhibiting a control algorithm.

omichel commented:
And yes, we should also modify the instructions to support Windows as well as described by @lukicdarkoo.
@BenjaminHug8: can you take care of this?
We will wait for the Windows instructions before testing again.

BenjaminHug8 commented:
> @BenjaminHug8: can you take care of this?

Of course, I am already working on it.

omichel commented on source/Tutorials/Simulators/Code/my_robot_driver.py:
```diff
@@ -0,0 +1,53 @@
+        command_left = 0
+        command_right = 0
+
+        if self.__target_twist.linear.x != 0.0 or self.__target_twist.linear.y != 0.0:
+            if self.__target_twist.linear.x <= 0.0:
+                if self.__target_twist.linear.y > 0.0:
+                    command_left = MAX_SPEED
+                    command_right = -MAX_SPEED
+                else:
+                    command_left = -MAX_SPEED
+                    command_right = MAX_SPEED
+            else:
+                vector_move_norm = math.sqrt(math.pow(self.__target_twist.linear.x, 2) + math.pow(self.__target_twist.linear.y, 2))
+                ratio_turn = self.__target_twist.linear.y / vector_move_norm
+                command_left = MAX_SPEED * clamp(1 + ratio_turn, 0, 1)
+                command_right = MAX_SPEED * clamp(1 - ratio_turn, 0, 1)
```
That seems wrong. The conversion from the twist message (desired linear and angular velocities) to motor speeds for a differential wheels robot is not correct. We should correct it and explain it a little bit in the text.

BenjaminHug8 replied:
Should be resolved with the `update conversion speed` commit.

omichel replied:
Thank you.
However, I believe it is better to use "counterclockwise" (American English) rather than anti-clockwise (British English).
Also, I wouldn't change the fact that the robot goes forward and bump into the wall. Because it is nice to see the problem and then come with a solution (use the sensors to avoid the obstacles).

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,400 @@
+      .. code-block:: console
+
+        # Install webots_ros2 and dependencies
+        cd \ros2_ws
```
In most of the other tutorials, this folder is named `dev_ws` instead. Only a few tutorials seems to use `ros2_ws` (after introducing it explicitly). I believe it is better to remain consistent with the majority of the tutorials:
```suggestion
        cd \dev_ws
```

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,400 @@
+
+      .. code-block:: console
+
+        cd ~/ros2_ws/src
```
```suggestion
        cd ~/dev_ws/src
```

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,400 @@
+
+      .. code-block:: console
+
+        cd \ros2_ws\src
```
```suggestion
        cd \dev_ws\src
```

BenjaminHug8 commented:
Shall we ask @lukicdarkoo and @joachimhgg to check again the tutorial now @omichel?

omichel commented:
Lets finalize the clean-up before.

BenjaminHug8 commented:
I removed unused images, the file .md containing the tutorial on `webots_ros2` to create a custom robot and I also changed some names in the duplicated files for the part 2 with sensors (e.g. `"My robot Webots"` into `"My robot"`).

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,400 @@
+
+Your robot should go forward and before hitting the wall it should turn clockwise.
+
+.. image:: Image/Robot_clock_wise.png
```
```suggestion
.. image:: Image/Robot_clockwise.png
```
And we have to rename the file as well.

omichel replied:
Sorry, it would be better with a verb to be consistent with other file names.
```suggestion
.. image:: Image/Robot_turning_clockwise.png
```

BenjaminHug8 replied:
Done.

omichel commented:
Thank you. I believe we can ask now to @lukicdarkoo and @joachimhgg to test it again carefully and report any problem they may spot. It would be nice if one of you could test it on Windows.

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,398 @@
+.. code-block:: console
+
+        cd my_package
+        mkdir launch worlds
```
This doesn't work on Windows.

BenjaminHug8 replied:
No problem on my side with `mkdir` in Visual Studio Command Prompt.

lukicdarkoo commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,398 @@
+
+4 Change the my_robot_driver.py file
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
Here we need an explanation of why we want to change the file. We should mention:
- that `webots_ros2_driver` automatically creates a ROS 2 interface for most sensors
- that one can extend the interface using C++ or Python plugins
- that the differential drive Webots/ROS plugin doesn't exist as it is a part of the `ros2_control` package, but we are creating this file for demonstration purposes

lukicdarkoo commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,398 @@
+    :lines: 14-16
+
+Then, the ROS node interacting with the simulated robot is created.
+This node, named ``driver``, is located in the ``webots_ros2_driver`` package.
```
Maybe, we should also explain to users that the `driver` node uses a custom protocol based on IPC and shared memory to communicate with a robot in the simulation. A little technical details might be interesting to users

lukicdarkoo commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,398 @@
+    :language: python
+    :lines: 18-25
+
+Finally, the two nodes are started in such a way that if the ``webots`` launcher node is shutdown, the other ``my_robot_driver`` node will be shutdown as well.
```
This sentence is a bit confusing. I would use one sentence to explain:
```
        webots,
        my_robot_driver,
```
and another one to explain:
```
launch.actions.RegisterEventHandler(
            event_handler=launch.event_handlers.OnProcessExit(
                target_action=webots,
                on_exit=[launch.actions.EmitEvent(event=launch.events.Shutdown())],
            )
        )
```
The second part is optional and it shutdowns all nodes once the Webots process dies.

lukicdarkoo commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,398 @@
+7 Modify the setup.py file
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Finally before you can start the launch file, you have to modify the setup.py file to include the extra files you added.
```
```suggestion
Finally, before you can start the launch file, you have to modify the `setup.py` file to include the extra files you added.
```

lukicdarkoo commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,398 @@
+Let's modify these two files to enable the sensors.
+Open ``package.xml`` and add the following content inside the ``<package format="3">`` tag:
+
+.. literalinclude:: Code/package_sensor.xml
```
```suggestion
.. literalinclude:: Code/package_sensor.xml
```
Maybe we can avoid explaining this line as well because it is a dependency of webots_ros2_driver

BenjaminHug8 replied:
Do you suggest to put it in the creation of the package?

In my opinion the user might not understand why there is this package when he is in the 1st part of the tutorial (without the sensors).
Currently the tutorial will use directly anything that has been setup, so the user can make links between the different files.

lukicdarkoo replied:
We just not include it in the tutorial. `sensor_msgs` comes with `ros-base`, so it should without explicitly specifying it

lukicdarkoo replied:
It is even a part of `ros-core`:
```
$ apt-rdepends ros-galactic-ros-core | grep sensor
Reading package lists... Done
Building dependency tree       
Reading state information... Done
  Depends: ros-galactic-sensor-msgs
ros-galactic-sensor-msgs
  Depends: ros-galactic-sensor-msgs
```

Don't get me wrong, if a package is going into production I would explicitly put `sensor_msgs` as a dependency. However, if we want to make the tutorial a little bit simpler then there is no strong need to mention it.

BenjaminHug8 replied:
Ok I will remove it then, thank you!

lukicdarkoo reviewed (changes requested):
A few more suggestions to make it simpler

BenjaminHug8 commented:
I changed the installation part to get only `webots_ros2_driver` but I kept the background on `webots_ros2` and added a note to install `webots_ros2`.

I think it is great for users to have this but it might be too "commercial" for ROS2 committee.

Let me know what you think about it.

omichel commented:
I committed a little polishing of the latest contributions which look good to me.

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,396 @@
+
+.. note::
+
+    If you want to install the whole ``webots_ros2`` package, follow these `instructions <https://github.com/cyberbotics/webots_ros2/wiki/Getting-Started`_.
```
broken markdown, also it is pointing to the same tutorial

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,396 @@
+    However, you could avoid the use of this Python plugin by using another ``webots_ros2`` sub-package named ``webots_ros2_control``, introducing a new dependency.
+    This other sub-package creates an interface with the ``ros2_control`` package that facilitates the control of a differential wheeled robot.
+
+Open ``my_package/my_robot_driver.py`` in your favorite editor and replace its contents with the following:
```
```suggestion
Open ``src/my_package/my_package/my_robot_driver.py`` in your favorite editor and replace its contents with the following:
```
to avoid confusion with my_package folders

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,396 @@
+11 Test the obstacle avoidance code
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Repeat the same commands as in tasks ``8`` to test your code.
```
```suggestion
Repeat the same commands as in tasks ``7`` to test your code.
```

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,396 @@
+
+        colcon build
+        source install/local_setup.bash
+        ros2 launch my_package robot_launch.py
```
do not work for me when I follow the step 7 again. The controller does not start. Also when I launch the command ros2 launch I get: 

```
WARNING: Target Webots version `R2021b` is not found, fallback to `R2021b revision 1`
[INFO] [webots-1]: process started with pid [201634]
[INFO] [driver-2]: process started with pid [201636]
[driver-2] Cannot open file: /tmp/webots-201634-vLpU51/WEBOTS_SERVER (retry count 1)
[driver-2] Cannot open file: /tmp/webots-201634-vLpU51/WEBOTS_SERVER (retry count 2)
```

BenjaminHug8 replied:
As discussed together, it seems you miss a part of the tutorial.
Was it easy to miss that part or unclear?

joachimhgg replied:
I wrote it in the last review

joachimhgg reviewed (commented):
Issue with the last part. Also do not forget to update the world.wbt with 2022a when available.

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,396 @@
+-------------
+
+It is recommended to understand basic ROS principles covered in the beginner :doc:`../../Tutorials`.
+In particular, :doc:`../Turtlesim/Introducing-Turtlesim`, :doc:`../Topics/Understanding-ROS2-Topics`, :doc:`../Workspace/Creating-A-Workspace`, :doc:`../Creating-Your-First-ROS2-Package` and :doc:`../Launch-Files/Creating-Launch-Files` are useful prerequisites.
```
No, but they give a flavor on how to work with a super simple, non-realistic, simulator. So, this might be helpful, although not strictly necessary.

omichel commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,396 @@
+
+      .. code-block:: console
+
+        sudo apt update
+        sudo apt-get install ros-{DISTRO}-webots-ros2-driver
```
In that case, I would rather remove the `-get` suffix in both cases because it is the old legacy version.

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -355,7 +355,7 @@ This will create an ``obstacle_avoider`` node that will be included in the ``Lau
 
-Repeat the same commands as in tasks ``8`` to test your code.
+Repeat the same first commands as in task ``7`` to launch the simulation.
 From a terminal in your ROS 2 workspace run:
```
```suggestion
As in task ``7`` to launch the simulation, from a terminal in your ROS 2 workspace run:
```
to not make the impression that we have to repeat the commands of task 7 before moving to the next task.

omichel replied:
(I just edited your proposal to make it a little bit simpler)

joachimhgg replied:
it could be nice to change the background of the code sections when they are related to explication

BenjaminHug8 replied:
It seems not so easy to do it. I do not find a way to change only specific `literalinclude`.
I tried to play with `:emphasize-lines:` but the background is also yellow so we can not see any change.

Furthermore it will be quite different from all the other tutorials as none of them is changing the default background color.
@omichel do you think I should spend more time on this?

omichel replied:
> @omichel do you think I should spend more time on this?

No.

lukicdarkoo commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,411 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You will need a world file containing a robot to launch your simulation.
+:download:`Download this world file <Code/my_world.wbt>` and move it inside ``my_package/worlds/``.
```
I am not sure whether it is a good idea, but we can provide a curl command for it. I saw they used it in the main ROS tutorial for Windows, so ROS assumes there is the curl command. However, the curl command throws an HTTPS-related error when called from the VS 2019 command prompt.

omichel replied:
I would avoid it if it throws errors in the console... And it seems the VS 2019 command prompt is often used with ROS 2, isn't it?

lukicdarkoo replied:
That's true, it is always used for compilation, so for this tutorial as well. I agree, it would be nice, but since we cannot find a clean solution it is better to avoid it.

lukicdarkoo commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,411 @@
+
+.. code-block:: console
+
+            source install/setup.bash
```
```suggestion
```
This is not needed as we don't use any custom messages

lukicdarkoo commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,411 @@
+.. code-block:: console
+
+            source install/setup.bash
+            ros2 topic pub /cmd_vel geometry_msgs/Twist  '{linear:  {x: 0.1, y: 0.0, z: 0.0}, angular: {x: 0.0, y: 0.0, z: 0.0}}'
```
```suggestion
            ros2 topic pub /cmd_vel geometry_msgs/Twist  "linear: { x: -0.1 }"
```
Single quotes don't work with Windows, also, it can be simplified

omichel commented:
@joachimhgg and @lukicdarkoo: please approve this PR if you believe it is fine for you.

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,410 @@
+
+.. code-block:: console
+
+            ros2 topic pub /cmd_vel geometry_msgs/Twist  "linear: { x: -0.1 }"
```
```suggestion
            ros2 topic pub /cmd_vel geometry_msgs/Twist  "linear: { x: 0.1 }"
```
To go forward

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,410 @@
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+You have to modify the URDF file in order to enable the sensors.
+In ``my_robot.urdf`` add the following content inside the ``<webots>`` tag:
```
It was not working for me because I replaced the content inside the `webots` tag and so remove the `plugin`. Through the tutorial, we only have to replace code or create new file. To avoid the confusion, I suggest either to specify that we need to keep plugin or add the plugin into the proposed code and change `add` by `replace`.

joachimhgg commented on source/Tutorials/Simulators/Setting-up-a-Robot-Simulation-Webots.rst:
```diff
@@ -0,0 +1,410 @@
+11 Test the obstacle avoidance code
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Repeat the same first commands as in task ``7`` to launch the simulation.
+From a terminal in your ROS 2 workspace run:
+
```
```suggestion
```

BenjaminHug8 commented:
I think the tutorial is ready now.
